### ISSUE TITLE
Lint not working in inline templates with custom component suffix

### ISSUE THREAD

**Issue by cyrilletuzi** [angular-eslint/angular-eslint#157]
First, thanks for the hard work! 👍🏻 

Using different component suffixes to differentiate component types (pure / dumb / presentational vs impure / page / ~ controller) is a widely used and good practice in Angular, which is officially supported by:
- the `component-class-suffix` from `codelizer`, which is already supported by `angular-eslint`
- the `--type` option in Angular CLI >= 9: `ng g component hello --type page`

But lint is not working with `angular-eslint` inside components using *inline* templates and a *different suffix* than `.component.ts`.

Here is a reproduction repo: [https://github.com/cyrilletuzi/ngeslintsuffixissue](https://github.com/cyrilletuzi/ngeslintsuffixissue)

Steps done:
1. `ng new ngeslint`
2. `ng add @angular-eslint/schematics`
3. `ng g @angular-eslint/schematics:add-config-to-project ngeslint`
4. adjusted the config to support `*.page.html/ts` both in `.eslintrc.json` and `angular.json`
5. created 4 components with inline or external template and default or custom suffix, with inverted banana box (`([ngModel])`)

Just run `npm run lint` (which was modified to use `angular-eslint`) and you'll see the issue is reported:
- in component with external template and default `.component.ts` suffix
- in component with inline template and default `.component.ts` suffix
- in component with external template and custom `.page.ts` suffix
- but NOT in component with inline template and custom `.page.ts` suffix

Same results when using VS Code ESLint extension.

**Comment by JamesHenry** [comment]
Yep, we try and skip a lot of unnecessary work pre and post-processing files that are not Components, so for now the processors skip anything that does not end in `.component.ts`.

Unfortunately what we really want is https://github.com/eslint/eslint/pull/12068 but it looks like the author abandoned it. I may try and pick that up myself at some point, because the concept as been approved by the ESLint Team by an RFC already (although obviously then the option will only be supported from that version of ESLint onwards).

For now I will go with "complaint driven development" and add explicit handling for `.page.ts`. I genuinely don't think there are a ton of different suffixes out there that people use and this is the first time even page (which I can understand being more common) has come up.

If we get more issues reported around different suffixes then we can evaluate things again.

**Comment by cyrilletuzi** [comment]
@JamesHenry Thanks for you answer.

Why is `.component.ts` hard-coded while the configuration exist in `.eslintrc.json` with `files: ['*.component.ts']`, and thus can be changed by the user to `files: ['*.component.ts', '*.page.ts']`? Can't it be taken into account like it is for external HTML files?

Note that I took `*.page.ts` as an example, but there are already other kind of common suffixes. For example:
- Material introduces dialogs, thus `*.dialog.ts`
- Ionic introduces modals, thus `*.modal.ts`

**Comment by JamesHenry** [comment]
You are correct that the user has that control over what files they run on, the trouble is it is easy for users to create non-performant lint configs.

So what this is about it about preventing the user from doing something very open like `files: ['*.ts']` or equivalent when wiring up the ESLint processor for inline-template extraction.

In that setup they are forcing us to do an extra parse of every single TypeScript file in their workspace and infer if there is a component definition within it.

In order to address this for now, I am going to implement handling for all of your suggested suffixes _and_ log an explicit warning in the case where the user has a config that does not match the expectations of the processor, referencing back to this comment.

**We can always add support for more suffixes if they come up**, and then hopefully contribute the `processOptions` to ESLint itself, which will remove this entire problem, as the user will be able to instruct the processor what to consider for their setup (and the schematics could auto generate that config).

**Comment by cyrilletuzi** [comment]
@JamesHenry 

> So what this is about it about preventing the user from doing something very open like files: ['*.ts'] or equivalent when wiring up the ESLint processor for inline-template extraction.

Can't we just check that the config is not just `*.ts` but `*.something.ts`?

If we stick to a fixed list, [here is the full list of special components](https://github.com/cyrilletuzi/vscode-angular-schematics/blob/d7e1bb587066b3ff2f9d2af73b2d69f7066fbbea/src/defaults.ts#L34) I already have identified in popular tools for Angular schematics extension for VS Code:
- `.page.ts` (common Angular practice + Ionic)
- `.dialog.ts` (Material + PrimeNG)
- `.snackbar.ts` (Material)
- `.bottomsheet.ts` (Material)
- `.modal.ts` (Ionic)
- `.popover.ts` (Ionic)

For the fixed list solution, I can take care of the PR if you want. For a dynamic one, I could too but I would need some indications about how to get the `.eslintrc.json` user config in [processors.ts](https://github.com/angular-eslint/angular-eslint/blob/9c2b5475d94c4d29eb4997b100c9af08351b722c/packages/eslint-plugin-template/src/processors.ts#L18).

**Comment by JamesHenry** [comment]
Again with that suggestion we would still be checking every Service, Provider, Module etc in an Angular project - Angular uses named extensions a lot.

We could obviously approach it a few different ways, but let's not overthink the solution for now, I have implemented your list (thank you for putting that together) and the warning that makes it clear what is happening in the uncommon case:

![image](https://user-images.githubusercontent.com/900523/95978268-d70b3180-0e11-11eb-8175-b57bbc16cf37.png)

**Comment by cyrilletuzi** [comment]
Thanks for the implementation! 👍🏻 

> Again with that suggestion we would still be checking every Service, Provider, Module etc in an Angular project - Angular uses named extensions a lot.

Not sure we understand each other. I'm not suggesting to try to extract inline templates from any file with a suffix (`.anything.ts`), but to use the configuration in `.eslintrc.json`:

```json
{
  "files": ["*.component.ts", "*.page.ts"],
  "extends": ["plugin:@angular-eslint/template/process-inline-templates"]
}
```

For sure, checking that the user doesn't just put `*.ts` in the config is a good thing, but why the config above would impact also services, modules, etc.?

**Comment by JamesHenry** [comment]
The config above wouldn't, and this is not about when users configure things correctly - it's about protecting against the opposite.

You need to consider the different bits of machinery at play - all of this revolves around what logic happens in the ESLint custom processor we are providing, specifically.

A custom processor has no control over what files it is invoked on - that is down to the ESLint config. The user controls the ESLint config, not us.

This means that a user can set up a config that gets them exactly what they want, as you have done above, BUT they can also set themselves up a config that shoots themselves in the foot.

In order to avoid the latter case, we have additional logic within the processor itself to bail out of work if it feels it has been invoked on a file it shouldn't have been.

Ideally the processor would accept configuration itself, but it currently doesn't in the latest version of ESLint (what I reference in this comment here: https://github.com/angular-eslint/angular-eslint/issues/157#issuecomment-708222158)

Hopefully that makes things a little clearer about the different pieces of ESLint that are involved

**Comment by JamesHenry** [comment]
@cyrilletuzi Just FYI I have removed the use of suffixes etc in PR #173 and it forms part of the breaking changes for v0.7.

To replace the suffix checking I have added some initial string based heuristics to work on the file contents to try and infer that it is worth even trying to parse the file and check for components.

It is already available for you or anyone else to use in `0.7.0-alpha.12`, published from that PR.
